src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -884,33 +884,21 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
         private void DecodeUnmanagedCallersOnlyAttribute(ref DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
         {
             Debug.Assert(arguments.AttributeSyntaxOpt != null);
-            if (!IsStatic || MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction))
-            {
-                // `UnmanagedCallersOnly` can only be applied to ordinary static methods or local functions.
-                arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, arguments.AttributeSyntaxOpt.Location);
-            }
 
             arguments.GetOrCreateData<MethodWellKnownAttributeData>().UnmanagedCallersOnlyAttributeData =
                 DecodeUnmanagedCallersOnlyAttributeData(arguments.Attribute, arguments.AttributeSyntaxOpt.Location, arguments.Diagnostics);
 
-            var (returnTypeSyntax, genericTypeParameters) = SyntaxNode switch
-            {
-                MethodDeclarationSyntax m => (m.ReturnType, m.TypeParameterList),
-                LocalFunctionStatementSyntax l => (l.ReturnType, l.TypeParameterList),
-                _ => default
-            };
+            bool reportedError = CheckAndReportValidUnmanagedCallersOnlyTarget(arguments.AttributeSyntaxOpt.Name.Location, arguments.Diagnostics);
 
-            if (returnTypeSyntax is null)
-            {
-                // We already issued an error for these above. We don't immediately bail out so we can error on invalid
-                // calling convention types as well, but erroring on parameter or return types will likely just be noise.
-                Debug.Assert(MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction));
-                return;
-            }
+            var returnTypeSyntax = this.ExtractReturnTypeSyntax();
 
-            if (isGenericMethod(this) || ContainingType.IsGenericType)
+            if (returnTypeSyntax is CompilationUnitSyntax && this is not SynthesizedSimpleProgramEntryPointSymbol)
```

Comment:
>returnTypeSyntax is CompilationUnitSyntax && this is not SynthesizedSimpleProgramEntryPointSymbol [](start = 16, length = 97)

Perhaps we should simply check if what we got back is ```CSharpSyntaxTree.Dummy.GetRoot()```? The current kind check and the SynthesizedSimpleProgramEntryPointSymbol check look too cryptic.